### PR TITLE
PETSc: Support MPItrampoline

### PR DESCRIPTION
**Pull request by eschnett**

### CODE REVIEW

**Comment by eschnett** [comment]
The build completes, but the BinaryBuilder check fails with:
```
MPItrampoline: Could not dlopen library "/agent/_work/1/s/P/PETSc/build/x86_64-linux-gnu-libgfortran5-mpi+mpitrampoline/mwMKFvzc/x86_64-linux-gnu-libgfortran5-cxx11-mpi+mpitrampoline/destdir/lib/petsc/double_real_Int32/lib/../../lib/libmpiwrapper.so"
MPItrampoline: dlerror: /agent/_work/1/s/P/PETSc/build/x86_64-linux-gnu-libgfortran5-mpi+mpitrampoline/mwMKFvzc/x86_64-linux-gnu-libgfortran5-cxx11-mpi+mpitrampoline/destdir/lib/petsc/double_real_Int32/lib/../../lib/libmpiwrapper.so: cannot open shared object file: No such file or directory
┌ Warning: lib/petsc/double_real_Int32/lib/libpetsc_double_real_Int32.so.3.16.5 cannot be dlopen()'ed
└ @ BinaryBuilder.Auditor /depot/packages/BinaryBuilder/wohhx/src/Auditor.jl:171
##[warning]lib/petsc/double_real_Int32/lib/libpetsc_double_real_Int32.so.3.16.5 cannot be dlopen()'ed
```

I assume the reason is the following:
- `libpetsc_double_real_Int32` is dlopened
- this causes its dependency `libmpitrampoline.so` to be loaded
- this runs startup code, which wants to open `libmpiwrapper.so`
- that library is not found, possibly because the location is wrong

**Comment by eschnett** [comment]
All is good, I solved the problem mentioned above by setting an environment variable.

File: P/PETSc/build_tarballs.jl [hunk 1]
```diff
@@ -1,4 +1,7 @@
 using BinaryBuilder
+using Base.BinaryPlatforms
+const YGGDRASIL_DIR = "../.."
+include(joinpath(YGGDRASIL_DIR, "platforms", "mpi.jl"))
 
 name = "PETSc"
 version = v"3.16.5"
```
Comment (giordano) [inline P/PETSc/build_tarballs.jl:7]: It's typically preferable adding new dependencies when changing version number.  There is a new upstream version now: 3.16.6

File: P/PETSc/bundled/patches/mpi-constants.patch [hunk 2]
```diff
@@ -0,0 +1,239 @@
+diff --git a/src/mat/impls/aij/mpi/superlu_dist/superlu_dist.c b/src/mat/impls/aij/mpi/superlu_dist/superlu_dist.c
```
Comment (giordano) [inline P/PETSc/bundled/patches/mpi-constants.patch:1]: Would it be possible to upstream this patch?  It's a largish, upgrading petsc would be complicated if we shipped invasive patches.

File: P/PETSc/build_tarballs.jl (hunk 1 above)
Comment (eschnett) [reply P/PETSc/build_tarballs.jl:7]: Done.

File: P/PETSc/bundled/patches/mpi-constants.patch (hunk 2 above)
Comment (eschnett) [reply P/PETSc/bundled/patches/mpi-constants.patch:1]: I will do that.

Is it okay to keep the patch for the time being?

Comment (giordano) [reply P/PETSc/bundled/patches/mpi-constants.patch:1]: Yes, but going forward we always prefer needing as few patches as possible

**Pull request merged**